proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use an attribute like `[SwiftError]` on a by-ref or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError<T>` on a by-ref or `out` parameter to indicate the error parameter.
+3. Use a special return type `SwiftReturn<T, TError>` to indicate the error parameter and combine it with the return value.
+4. Use special helper functions `Marshal.Get/SetLastSwiftError` to get and set the last Swift error. Our various compilers will automatically emit a call to `SetLastSwiftError` from Swift functions and emit a call to `GetLastSwiftError` in the epilog of `UnmanagedCallersOnly` methods. The projection generation will emit calls to `Marshal.Get/SetLastSwiftError` to get and set these values to and from exceptions.
+
+We have a prototype that uses the first option; however, using an attribute has the same limitations as the attribute option for the self register. As a result, we should consider an alternative option.
+
+Options 2 and 3 have similar characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Option 2 is likely cheaper, as the argument with `ref/out SwiftError<T>` type can easily be associated with the error register. With Option 2 though, we'd have to determine the behavior if multiple parameters used the `SwiftError` type (likely throw an exception). Option 3 provides a more functional-programming-style API that is more intuitive, but likely more expensive to implement.
+
+Option 4 provides an alternative approach. As the "Error Register" is always register-sized, we can use runtime-supported helper functions to stash away and retrieve the error register. Unlike options 2 or 3, we don't need to do any signature manipulation as the concept of "this Swift call can return an error" is not represented in the signature. Responsibility to convert the returned error value into a .NET type, such as an exception, would fall to the projection tooling. Since option 4 does not express whether or not the target function throws at the signature level, the JIT/AOT compilers would always need to emit the calls to the helpers when compiling calls to and from Swift code. If the projection of Swift types into .NET will always use exception and not pass Swift errors as the error types directly, then Option 4 reduces the design burden on the runtime teams by removing the type that would only be used at the lowest level of the Swift/.NET interop.
```

Comment:
Also, this option leaves performance on the table.

Reply:
While I don't think this is a concern today, it would break if one day there's an architecture where the presence of the error return changes how other registers are used.

Reply:
Stashing away and retrieving the error later needs to be thread safe.